File: csharp/Tdcv2/Packs/PackSource.cs
using System.Reflection;

namespace Tdcv2.Packs;

/// <summary>
/// Where pack files come from.
/// </summary>
/// <remarks>
/// A folder on disk, or several searched in order. That is how this repository's tests read the same
/// packs the other implementations read, and how a project adds its own — a pack downloaded into a
/// project belongs to the project, not to whichever runtime happens to read it.
/// </remarks>
public interface IPackSource
{
    /// <summary>Whether a directory entry is repository noise rather than a pack.</summary>
    /// <remarks>
    /// This runs over directory names too, so it must not look at extensions — that is
    /// <see cref="IsPackFile"/>'s job. It is the same short list <c>load.ts</c> keeps.
    /// </remarks>
    static bool IsIgnoredEntry(string name)
    {
        if (name == "_locale.json" || name.StartsWith('.'))
        {
            return true;
        }

        int dot = name.LastIndexOf('.');
        string stem = (dot > 0 ? name[..dot] : name).ToLowerInvariant();
        return stem is "readme" or "license" or "changelog";
    }

    /// <summary>Whether a file name is a pack file rather than metadata sitting beside one.</summary>
    /// <remarks>
    /// The two extensions a pack FILE can carry: <c>.txt</c> for a list or a generator written in
    /// a header, <c>.tdc</c> for a generator written as a config. This used to allow everything —
    /// the spec said a data file's extension is ignored, written before a pack carried anything
    /// but data. Once <c>DATE_LOCALE.json</c> arrived, fifteen locales silently grew an address
    /// (<c>bn.DATE_LOCALE</c>, <c>hu.DATE_LOCALE</c>, …) whose values were the lines of the JSON
    /// source: <c>{</c>, <c>"months": [</c>, <c>"Január",</c>. No diagnostic, because nothing was
    /// malformed — a file was read as a list, which is what the loader was told to do with any
    /// file it found. An allowlist rather than a <c>.json</c> denylist, so the next metadata file
    /// to be added need not remember to come here first.
    /// </remarks>
    static bool IsPackFile(string name)
    {
        int dot = name.LastIndexOf('.');
        if (dot <= 0)
        {
            return false;
        }

        string extension = name[dot..].ToLowerInvariant();
        return extension is ".txt" or ".tdc";
    }

    /// <summary>Whether a pack exists at this path, e.g. <c>en/person/lastName.txt</c>.</summary>
    bool Has(string relativePath);

    IReadOnlyList<string> ReadLines(string relativePath);

    /// <summary>Every pack file this source can serve, as relative paths.</summary>
    /// <remarks>
    /// The address index is built from this. A file's header may place it at an address that has
    /// nothing to do with its path, and the only way to know is to look inside every file — which
    /// is why this exists and why the scan is deferred until a path lookup has already missed.
    /// </remarks>
    IReadOnlyList<string> ListFiles();

    /// <summary>
    /// Whether a top-level folder by this name exists — <c>en</c>, <c>usa</c>, <c>common</c>.
    /// </summary>
    /// <remarks>
    /// This is how a dotted address is judged absolute or relative: <c>usa.ssn</c> names a country
    /// pack directly, while <c>person.lastName</c> is relative to the active locale.
    /// </remarks>
    bool HasTopLevel(string name);

    /// <summary>
    /// Whether <c>countries/&lt;name&gt;</c> exists.
    /// </summary>
    /// <remarks>
    /// The <c>countries/</c> folder is a physical grouping, not part of an address: a pack at
    /// <c>countries/usa/finance/aba_routing.txt</c> is addressed as
    /// <c>usa.finance.aba_routing</c>. Without this, every country pack would look like a relative
    /// address and be searched for under the active locale, where it is not.
    /// </remarks>
    bool HasCountry(string name);

    /// <summary>Where a file physically is, for a complaint that has to name it.</summary>
    /// <remarks>
    /// A relative path is ambiguous the moment two roots are layered, and "which file did you
    /// mean" is the one thing such a message must not leave open. Packs compiled into the
    /// assembly have no path at all, which is why this can answer null.
    /// </remarks>
    string? Locate(string relativePath) => null;
}

/// <summary>One folder of packs.</summary>
public sealed class DirectorySource : IPackSource
{
    private readonly string _root;

    public DirectorySource(string root) => _root = root;

    private string Resolve(string relativePath) =>
        Path.Combine(_root, relativePath.Replace('/', Path.DirectorySeparatorChar));

    public bool Has(string relativePath) => File.Exists(Resolve(relativePath));

    public IReadOnlyList<string> ReadLines(string relativePath) =>
        File.ReadAllLines(Resolve(relativePath));

    public IReadOnlyList<string> ListFiles()
    {
        if (!Directory.Exists(_root))
        {
            return Array.Empty<string>();
        }

        string[] found = Directory.GetFiles(_root, "*", SearchOption.AllDirectories);
        var relative = new List<string>(found.Length);
        foreach (string path in found)
        {
            string rel = Path.GetRelativePath(_root, path).Replace(Path.DirectorySeparatorChar, '/');
            string[] parts = rel.Split('/');
            if (parts.Any(IPackSource.IsIgnoredEntry) || !IPackSource.IsPackFile(parts[^1]))
            {
                continue;
            }

            relative.Add(rel);
        }

        relative.Sort(StringComparer.Ordinal);
        return relative;
    }

    public bool HasTopLevel(string name) => Directory.Exists(Path.Combine(_root, name));

    public bool HasCountry(string name) =>
        Directory.Exists(Path.Combine(_root, "countries", name));

    public string? Locate(string relativePath)
    {
        string path = Path.Combine(_root, relativePath.Replace('/', Path.DirectorySeparatorChar));
        return File.Exists(path) ? path : null;
    }

    public override string ToString() => _root;
}

/// <summary>
/// Several sources searched in order, the last one winning.
/// </summary>
/// <remarks>
/// This is what lets a project's own packs shadow the bundled ones without replacing them: a
/// downloaded or hand-written pack at the same address is found first, and everything else still
/// resolves. The order is low priority to high, so the search runs backwards.
/// </remarks>
public sealed class LayeredSource : IPackSource
{
    private readonly IReadOnlyList<IPackSource> _sources;

    public LayeredSource(IReadOnlyList<IPackSource> sources) => _sources = sources.ToArray();

    private IPackSource? Owning(string relativePath)
    {
        for (int i = _sources.Count - 1; i >= 0; i--)
        {
            if (_sources[i].Has(relativePath))
            {
                return _sources[i];
            }
        }

        return null;
    }

    public bool Has(string relativePath) => Owning(relativePath) is not null;

    public IReadOnlyList<string> ReadLines(string relativePath) =>
        Owning(relativePath)?.ReadLines(relativePath)
        ?? throw new FileNotFoundException($"no pack source holds \"{relativePath}\"");

    public IReadOnlyList<string> ListFiles()
    {
        var out_ = new List<string>();
        var seen = new HashSet<string>(StringComparer.Ordinal);
        foreach (IPackSource source in _sources)
        {
            foreach (string path in source.ListFiles())
            {
                if (seen.Add(path))
                {
                    out_.Add(path);
                }
            }
        }

        return out_;
    }

    public bool HasTopLevel(string name) => _sources.Any(s => s.HasTopLevel(name));

    public bool HasCountry(string name) => _sources.Any(s => s.HasCountry(name));

    public string? Locate(string relativePath) => Owning(relativePath)?.Locate(relativePath);

    public override string ToString() => string.Join(", ", _sources);
}

/// <summary>
/// The starter packs compiled into the assembly.
/// </summary>
/// <remarks>
/// <para>
/// A NuGet package is a zip with an assembly in it and nothing above it, so looking beside the
/// assembly and then walking up for <c>data/packs</c> — which is how <see cref="DataPacks.Discover"/>
/// finds the repository's copy — cannot work in <c>~/.nuget/packages</c>. Embedding is the only
/// shape that survives every way a consumer can build: a plain library reference, a single-file
/// publish, a trimmed one.
/// </para>
/// <para>
/// Empty in a checkout, where the real <c>data/packs</c> is on disk and reading it from there is
/// what keeps all five implementations looking at the same bytes. <see cref="IsEmpty"/> is how a
/// caller tells the two apart.
/// </para>
/// <para>
/// Unlike the jar, this needs no index file: .NET can list its own resources.
/// </para>
/// </remarks>
public sealed class EmbeddedSource : IPackSource
{
    /// <summary>The resource-name prefix the .csproj gives every embedded pack.</summary>
    private const string Prefix = "tdc/packs/";

    private static readonly Assembly Owner = typeof(EmbeddedSource).Assembly;

    private static readonly IReadOnlyList<string> Paths = Owner
        .GetManifestResourceNames()
        .Where(n => n.StartsWith(Prefix, StringComparison.Ordinal))
        .Select(n => n[Prefix.Length..])
        .OrderBy(n => n, StringComparer.Ordinal)
        .ToArray();

    /// <summary>Whether anything was embedded — false in a checkout.</summary>
    public static bool IsEmpty => Paths.Count == 0;

    public bool Has(string relativePath) => Paths.Contains(relativePath);

    public IReadOnlyList<string> ReadLines(string relativePath)
    {
        using Stream? stream = Owner.GetManifestResourceStream(Prefix + relativePath);
        if (stream is null)
        {
            return Array.Empty<string>();
        }

        // Split the same way DirectorySource does, so an embedded pack and a pack on disk give
        // the identical list of values — including the trailing-newline rule.
        using var reader = new StreamReader(stream);
        var lines = new List<string>();
        while (reader.ReadLine() is { } line)
        {
            lines.Add(line);
        }

        return lines;
    }

    public IReadOnlyList<string> ListFiles() => Paths;

    public bool HasTopLevel(string name) =>
        Paths.Any(p => p.StartsWith(name + "/", StringComparison.Ordinal));

    public bool HasCountry(string name) =>
        Paths.Any(p => p.StartsWith("countries/" + name + "/", StringComparison.Ordinal));

    public string Describe() => $"the {Paths.Count} packs built into this assembly";
}

File: java/src/main/java/io/github/nickliapin/tdc/packs/PackSource.java
package io.github.nickliapin.tdc.packs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Where pack files come from.
 *
 * <p>Two places, and both have to work. The library ships every pack inside its own jar so that
 * {@code person.male.firstName} resolves with no configuration at all — a library that needed a
 * data directory pointed at it before it could produce a single name would be a poor library.
 * The other place is a folder on disk, which is how this repository's tests read the same packs
 * the other implementations read, and how a user adds their own.
 */
public interface PackSource {

  /**
   * Whether a directory entry is repository noise rather than a pack.
   *
   * <p>The reference ignores a file's extension entirely when it turns a path into an address, so
   * a name is the only thing that can keep a locale manifest, a README or a {@code .DS_Store} out
   * of the registry — and this is the same short list {@code load.ts} keeps. Everything else is
   * data, whatever it is called.
   */
  static boolean isIgnoredEntry(String name) {
    if (name.equals("_locale.json") || name.startsWith(".")) {
      return true;
    }
    int dot = name.lastIndexOf('.');
    String stem = dot > 0 ? name.substring(0, dot) : name;
    stem = stem.toLowerCase(java.util.Locale.ROOT);
    return stem.equals("readme") || stem.equals("license") || stem.equals("changelog");
  }

  /** Whether a pack exists at this path, e.g. {@code en/person/lastName.txt}. */
  boolean has(String relativePath);

  List<String> readLines(String relativePath);

  /**
   * Every pack file this source can serve, as relative paths.
   *
   * <p>The address index is built from this. A file's header may place it at an address that has
   * nothing to do with its path, and the only way to know is to look inside every file — which is
   * why this exists and why the scan is deferred until a path lookup has already missed.
   */
  List<String> listFiles();

  /**
   * Whether a top-level folder by this name exists — {@code en}, {@code usa}, {@code common}.
   *
   * <p>This is how a dotted address is judged absolute or relative: {@code usa.ssn} names a
   * country pack directly, while {@code person.lastName} is relative to the active locale.
   */
  boolean hasTopLevel(String name);

  /**
   * Whether {@code countries/<name>} exists.
   *
   * <p>The {@code countries/} folder is a physical grouping, not part of an address: a pack at
   * {@code countries/usa/finance/aba_routing.txt} is addressed as {@code usa.finance.aba_routing}.
   * Without this, every country pack would look like a relative address and be searched for
   * under the active locale, where it is not.
   */
  boolean hasCountry(String name);

  /**
   * Where a file physically is, for a complaint that has to name it.
   *
   * <p>A relative path is ambiguous the moment two roots are layered, and "which file did you
   * mean" is the one thing such a message must not leave open. Packs read from inside the jar
   * have no path at all, which is why this can answer {@code null}.
   */
  default String locate(String relativePath) {
    return null;
  }

  /**
   * Several sources searched in order, the last one winning.
   *
   * <p>This is what lets a project's own packs shadow the bundled ones without replacing them: a
   * downloaded or hand-written pack at the same address is found first, and everything else still
   * resolves. The order is low priority to high, so the search runs backwards.
   */
  final class Layered implements PackSource {
    private final List<PackSource> sources;

    public Layered(List<PackSource> sources) {
      this.sources = List.copyOf(sources);
    }

    private PackSource owning(String relativePath) {
      for (int i = sources.size() - 1; i >= 0; i--) {
        if (sources.get(i).has(relativePath)) {
          return sources.get(i);
        }
      }
      return null;
    }

    @Override
    public boolean has(String relativePath) {
      return owning(relativePath) != null;
    }

    @Override
    public List<String> readLines(String relativePath) {
      PackSource source = owning(relativePath);
      if (source == null) {
        throw new IllegalStateException("no pack at " + relativePath);
      }
      return source.readLines(relativePath);
    }

    @Override
    public List<String> listFiles() {
      List<String> out = new ArrayList<>();
      Set<String> seen = new HashSet<>();
      for (PackSource source : sources) {
        for (String path : source.listFiles()) {
          if (seen.add(path)) {
            out.add(path);
          }
        }
      }
      return out;
    }

    @Override
    public boolean hasTopLevel(String name) {
      for (PackSource source : sources) {
        if (source.hasTopLevel(name)) {
          return true;
        }
      }
      return false;
    }

    @Override
    public boolean hasCountry(String name) {
      for (PackSource source : sources) {
        if (source.hasCountry(name)) {
          return true;
        }
      }
      return false;
    }

    @Override
    public String locate(String relativePath) {
      PackSource source = owning(relativePath);
      return source == null ? null : source.locate(relativePath);
    }
  }

  /** Packs from a directory. */
  final class Directory implements PackSource {
    private final Path root;

    public Directory(Path root) {
      this.root = root;
    }

    @Override
    public boolean has(String relativePath) {
      return Files.exists(root.resolve(relativePath));
    }

    @Override
    public List<String> readLines(String relativePath) {
      try {
        return Files.readAllLines(root.resolve(relativePath), StandardCharsets.UTF_8);
      } catch (IOException e) {
        throw new UncheckedIOException("cannot read pack " + root.resolve(relativePath), e);
      }
    }

    @Override
    public List<String> listFiles() {
      if (!Files.isDirectory(root)) {
        return List.of();
      }
      try (java.util.stream.Stream<Path> walk = Files.walk(root)) {
        return walk.filter(Files::isRegularFile)
            .map(path -> root.relativize(path).toString().replace('\\', '/'))
            .filter(path -> java.util.Arrays.stream(path.split("/")).noneMatch(
                PackSource::isIgnoredEntry))
            .sorted()
            .toList();
      } catch (IOException e) {
        throw new UncheckedIOException("cannot list packs under " + root, e);
      }
    }

    @Override
    public boolean hasTopLevel(String name) {
      return Files.isDirectory(root.resolve(name));
    }

    @Override
    public boolean hasCountry(String name) {
      return Files.isDirectory(root.resolve("countries").resolve(name));
    }

    @Override
    public String locate(String relativePath) {
      Path path = root.resolve(relativePath);
      return Files.isRegularFile(path) ? path.toString() : null;
    }

    @Override
    public String toString() {
      return root.toString();
    }
  }

  /** Packs from inside the jar. */
  final class Classpath implements PackSource {
    /** Where the build puts them. */
    static final String PREFIX = "tdc/packs/";

    private final Set<String> index;
    private final Set<String> topLevel;
    private final Set<String> countries;

    public Classpath() {
      // A jar has no directories to list, so the build writes down what it packed. Probing for
      // each candidate instead would make "does the locale `sv` exist?" unanswerable — the only
      // way to find out would be to guess a filename inside it.
      this.index = new HashSet<>(readIndex());
      this.topLevel = new HashSet<>();
      this.countries = new HashSet<>();
      for (String path : index) {
        int slash = path.indexOf('/');
        if (slash > 0) {
          topLevel.add(path.substring(0, slash));
        }
        if (path.startsWith("countries/")) {
          int next = path.indexOf('/', "countries/".length());
          if (next > 0) {
            countries.add(path.substring("countries/".length(), next));
          }
        }
      }
    }

    /**
     * Every pack index on the classpath, merged.
     *
     * <p>Plural on purpose. Java already has a package manager, so the natural way to add packs
     * is another dependency — a jar carrying its own {@code tdc/packs/} tree and its own index.
     * Reading only the first index would make every such jar invisible, and the failure would
     * look like a missing pack rather than a missing feature.
     */
    /**
     * The loader to ask.
     *
     * <p>The thread's context loader when there is one, because that is the application's view of
     * the classpath — in a container or a framework's launcher, the jar this class came from sees
     * a narrower one, and a pack the application added would be invisible from here. Falls back
     * to this class's own loader, which is right for a plain command line and for tests.
     */
    private static ClassLoader loader() {
      ClassLoader context = Thread.currentThread().getContextClassLoader();
      return context != null ? context : Classpath.class.getClassLoader();
    }

    private static List<String> readIndex() {
      List<String> out = new ArrayList<>();
      try {
        java.util.Enumeration<java.net.URL> found = loader().getResources(PREFIX + "index.txt");
        while (found.hasMoreElements()) {
          try (InputStream in = found.nextElement().openStream()) {
            out.addAll(readAll(in));
          }
        }
      } catch (IOException e) {
        throw new UncheckedIOException("cannot read the bundled pack index", e);
      }
      // No index at all means a jar built without packs. Every lookup then fails by name, which
      // is a far better outcome than silently producing rows with empty fields.
      return out;
    }

    @Override
    public boolean has(String relativePath) {
      return index.contains(relativePath);
    }

    @Override
    public List<String> readLines(String relativePath) {
      InputStream in =
          loader().getResourceAsStream(PREFIX + relativePath);
      if (in == null) {
        throw new IllegalArgumentException("bundled pack is missing: " + relativePath);
      }
      return readAll(in);
    }

    @Override
    public List<String> listFiles() {
      List<String> out = new ArrayList<>(index);
      out.sort(null);
      return out;
    }

    @Override
    public boolean hasTopLevel(String name) {
      return topLevel.contains(name);
    }

    @Override
    public boolean hasCountry(String name) {
      return countries.contains(name);
    }

    @Override
    public String toString() {
      return "bundled packs (" + index.size() + " files)";
    }
  }

  private static List<String> readAll(InputStream in) {
    List<String> out = new ArrayList<>();
    try (BufferedReader reader =
        new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8))) {
      String line;
      while ((line = reader.readLine()) != null) {
        out.add(line);
      }
    } catch (IOException e) {
      throw new UncheckedIOException("cannot read a bundled pack", e);
    }
    return out;
  }
}
